src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -0,0 +1,958 @@
+
+        # change view [batch_size, num_heads, q_length, head_dim]
+        context_layer = context_layer.view(*output_size)
+
+        # [batchs_size, num_heads, q_length, head_dim] --> [q_length, batch_size, num_heads, head_dim]
+        context_layer = context_layer.permute(2, 0, 1, 3).contiguous()
+
+        # [q_length, batch_size, num_heads, head_dim] --> [q_length, batch_size, hidden_size]
+        new_context_layer_shape = context_layer.size()[:-2] + (self.hidden_size,)
+
+        context_layer = context_layer.view(*new_context_layer_shape)
+
+        # Output. [q_length, batch_size, hidden_size]
+
+        # aggregate results across tp ranks. See here: https://github.com/pytorch/pytorch/issues/76232
+        if self.pretraining_tp > 1 and self.slow_but_exact:
+            slices = context_layer.shape[-1] / self.pretraining_tp
+            output_tensor = torch.zeros_like(context_layer)
+            for i in range(self.pretraining_tp):
+                output_tensor = output_tensor + nn.functional.linear(
+                    context_layer[:, :, int(i * slices) : int((i + 1) * slices)],
+                    self.dense.weight[:, int(i * slices) : int((i + 1) * slices)],
+                )
+        else:
+            output_tensor = nn.functional.linear(context_layer, self.dense.weight)
+
+        output_tensor = output_tensor
+        output_bias = self.dense.bias
+
+        output = output_tensor.transpose(1, 0)
+        outputs = (output, present)
+        if output_attentions:
+            outputs += (attention_probs,)
+
+        return outputs, output_bias
+
+
+class BloomMLP(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        hidden_size = config.hidden_size
+
+        self.pretraining_tp = config.pretraining_tp
+        self.slow_but_exact = config.slow_but_exact
+        self.activation_func = bias_gelu_impl
+        self.dense_h_to_4h = nn.Linear(hidden_size, 4 * hidden_size)
+        self.dense_4h_to_h = nn.Linear(4 * hidden_size, hidden_size)
+
+    def forward(self, hidden_states):
+        input_ = hidden_states
+
+        hidden_states = self.activation_func(
+            nn.functional.linear(hidden_states, self.dense_h_to_4h.weight), self.dense_h_to_4h.bias
+        )
+
+        if self.pretraining_tp > 1 and self.slow_but_exact:
+            intermediate_output = torch.zeros_like(input_)
+            slices = self.dense_4h_to_h.weight.shape[-1] / self.pretraining_tp
+            for i in range(self.pretraining_tp):
+                intermediate_output = intermediate_output + nn.functional.linear(
+                    hidden_states[:, :, int(i * slices) : int((i + 1) * slices)],
+                    self.dense_4h_to_h.weight[:, int(i * slices) : int((i + 1) * slices)],
+                )
+        else:
+            intermediate_output = nn.functional.linear(hidden_states, self.dense_4h_to_h.weight)
+
+        output = intermediate_output
+        output_bias = self.dense_4h_to_h.bias
```

Comment:
I assume we don't do `output + output_bias` here because it's more efficient to do it later all at once? 
Does it make that big of a difference? 
In my opinion, it'd be much more readable if we do the addition here and just pass `hidden_states` as a single tensor (but ok for me to leave as is if it gives a big speed up)